arrow-libs/fx/arrow-fx-resilience/src/commonMain/kotlin/arrow/fx/resilience/Schedule.kt

```diff
@@ -1,1091 +1,500 @@
+      when (val decision = self(input)) {
+        is Done -> Done(other(decision.output).output)
+        is Continue -> when (val decision2 = other(decision.output)) {
+          is Done -> Done(decision2.output)
+          is Continue -> Continue(decision2.output, decision.delay + decision2.delay) {
+            loop(it, decision.next, decision2.next)
+          }
+        }
+      }
 
-  /**
-   * Executes one schedule after the other. When the first schedule ends, it continues with the second.
-   */
-  public abstract infix fun <A : Input, B> andThen(other: Schedule<A, B>): Schedule<A, Either<Output, B>>
+    return Schedule { input -> loop(input, step, other.step) }
+  }
 
-  /**
-   * Changes the delay of a resulting [Decision] based on the [Output] and the produced delay.
-   *
-   */
-  public abstract fun modify(f: suspend (Output, Duration) -> Duration): Schedule<Input, Output>
+  /** Runs `this` [Schedule] _while_ the [predicate] of [Input] and [Output] returns `false`. */
+  public fun doWhile(predicate: suspend (Input, Output) -> Boolean): Schedule<Input, Output> {
+    suspend fun loop(input: Input, self: Next<Input, Output>): Decision<Input, Output> =
+      when (val decision = self(input)) {
+        is Continue ->
+          if (predicate(input, decision.output)) Continue(decision.output, decision.delay) { loop(it, decision.next) }
+          else Done(decision.output)
 
-  @ExperimentalTime
-  @Deprecated(NanosDeprecation,
-    ReplaceWith("modify { output, d -> f(output, d.toDouble(DurationUnit.NANOSECONDS)).nanoseconds }",
-      "kotlin.time.DurationUnit", "kotlin.time.Duration.Companion.nanoseconds"))
-  public fun modifyNanos(f: suspend (Output, Double) -> Double): Schedule<Input, Output> =
-    modify { output, d -> f(output, d.toDouble(NANOSECONDS)).nanoseconds }
+        is Done -> decision
+      }
 
-  /**
-   * Runs an effectful handler on every input. Does not alter the decision.
-   */
-  public abstract fun logInput(f: suspend (input: Input) -> Unit): Schedule<Input, Output>
+    return Schedule { input -> loop(input, step) }
+  }
 
   /**
-   * Runs an effectful handler on every output. Does not alter the decision.
+   * Runs the [Schedule] _until_ the [predicate] of [Input] and [Output] returns true.
+   * Inverse version of [doWhile].
    */
-  public abstract fun logOutput(f: suspend (output: Output) -> Unit): Schedule<Input, Output>
+  public fun doUntil(predicate: suspend (input: Input, output: Output) -> Boolean): Schedule<Input, Output> =
+    doWhile { input, output -> !predicate(input, output) }
 
   /**
-   * Accumulates the results of a schedule by folding over them effectfully.
+   * Adds a logging action to the [Schedule].
    */
-  public abstract fun <C> foldLazy(
-    initial: suspend () -> C,
-    f: suspend (acc: C, output: Output) -> C
-  ): Schedule<Input, C>
+  public fun log(action: suspend (input: Input, output: Output) -> Unit): Schedule<Input, Output> =
```

Comment:
I was thinking about that too, but if we go with the `typealias` change then this would conflict with `also` from Kotlin. Since Kotlin will prefer `also` from Kotlin Std (requires no import) over our extension function.

While I am a big fan of naming things like the Kotlin Std. Naming things the same as Kotlin Std that are extension functions over `A` are a bit tricky from user perspective.